clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	goerrors "github.com/go-errors/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+	"sync"
+	"time"
+)
+
+const (
+	apiAvailabilityName = "ApiAvailability"
+)
+
+func init() {
+	if err := measurement.Register(apiAvailabilityName, createApiAvailabilityMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", apiAvailabilityName, err)
+	}
+}
+
+func createApiAvailabilityMeasurement() measurement.Measurement {
+	return &apiAvailabilityMeasurement{}
+}
+
+type apiAvailabilityOutput struct {
+	ClusterMetrics             bool          `json:"clusterMetrics"`
+	HostIp                     string        `json:"hostIp"`
+	AvailabilityPercentage     float32       `json:"availabilityPercentage"`
+	LongestUnavailableDuration time.Duration `json:"longestUnavailableDuration"`
+}
+
+type apiAvailabilityMetrics struct {
+	numProbes                  int
+	numSuccess                 int
+	numFailures                int
+	maxConsecutiveFailedProbes int
+	consecutiveFailedProbes    int
+}
+
+type apiAvailabilityMeasurement struct {
+	isRunning           bool
+	stopCh              chan struct{}
+	hosts               []string
+	clusterLevelMetrics *apiAvailabilityMetrics
+	hostLevelMetrics    map[string]*apiAvailabilityMetrics
+	wg                  sync.WaitGroup
+}
+
+func (a *apiAvailabilityMeasurement) updateMasterAvailabilityMetrics(c clientset.Interface, config *measurement.MeasurementConfig) {
+	for _, host := range a.hosts {
+		// update the number of probes
+		a.hostLevelMetrics[host].numProbes++
+		// SSH and check the health of the host
+		command := fmt.Sprintf("curl -s -k %slocalhost:%v/healthz", "https://", 443)
+		_, err := measurementutil.SSH(command, host+":22", "")
+		if err != nil {
```

Comment:
Does it really do what you want? If the command fails with non-zero error code, is the err always returned?

Reply:
Do you mean that err won't capture failure of command with non-zero error code?

Reply:
Yes - I think so (which is kind-of by-design):
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/ssh.go#L65
https://github.com/kubernetes/perf-tests/blob/2eeb004780c3f0ec99d0e03e6e44261e03017d6f/clusterloader2/vendor/k8s.io/kubernetes/pkg/ssh/ssh.go#L217